### PR TITLE
New functions - mixed context overloads

### PR DESCRIPTION
**Pull request by tomblind**
This update addresses issues with function overloads that mix signatures with and without context:
```ts
declare class Foo {
    method(s: string): string; // Has context
    method(this: void, s: string, n: number): string; // No context
}
```

Now, calling the overload will correctly use dot or colon based on the resolved signature:
```ts
declare const foo: Foo;
foo.method("foo");
foo.method("bar", 17);
```
```lua
foo:method("foo"); -- Uses colon
foo.method("bar", 17); -- Uses dot
```

Additionally, assignments to and from function types that have mixed overloads are now banned. Properly validating these assignments would require manually signature matching the overloads and would add a lot of complexity for a fairly fringe-y case. The error advises using only overloads that are all functions or all methods, but not both.
```ts
interface Func { (s: string, n: number): string };
let fn: Func = foo.method; // Error: Unsupported assignment of mixed function/method overload. All overloads should either be functions or methods.
```

### CODE REVIEW

File: src/Errors.ts [hunk 1]
```diff
@@ -84,4 +84,16 @@ export class TSTLErrors {
...
     }
+
+    public static UnsupportedOverloadAssignment = (node: ts.Node, name?: string) => {
+        if (name) {
+            return new TranspileError(`Unsupported assignment of mixed function/method overload to "${name}". `
+                                      + `All overloads should either be functions or methods.`,
+                                      node);
+        } else {
+            return new TranspileError(`Unsupported assignment of mixed function/method overload. `
+                                      + `All overloads should either be functions or methods.`,
```
Comment (Perryvw) [inline src/Errors.ts]: This error is a little missleading, how about rephrasing to 'Overloads should either all be methods, or all be functions.'

File: src/TSHelper.ts [hunk 2]
```diff
@@ -249,37 +256,43 @@ export class TSHelper {
...
             && !(ts.getCombinedModifierFlags(sigDecl) & ts.ModifierFlags.Static)) {
             // Non-static method
-            return true;
+            return ContextType.NonVoid;
         }
         if ((ts.isPropertySignature(sigDecl.parent) || ts.isPropertyDeclaration(sigDecl.parent))
             && !(ts.getCombinedModifierFlags(sigDecl.parent) & ts.ModifierFlags.Static)) {
             // Non-static lambda property
-            return true;
+            return ContextType.NonVoid;
```
Comment (Perryvw) [inline src/TSHelper.ts]: I thought lambda properties are currently transpiled without context, would that not mean void type here, or do we want to change this?

File: src/Transpiler.ts [hunk 3]
```diff
@@ -1269,8 +1269,9 @@ export abstract class LuaTranspiler {
                 params = this.transpileArguments(node.arguments, sig);
                 return `(rawget(${expr}, ${params} )~=nil)`;
             } else {
-                const type = this.checker.getTypeAtLocation(node.expression);
-                const op = tsHelper.isFunctionWithContext(type, this.checker) ? ":" : ".";
+                const sigDecl = sig.getDeclaration();
+                const op = !sigDecl || tsHelper.getDeclarationContextType(sigDecl, this.checker) !== ContextType.Void
```
Comment (Perryvw) [inline src/Transpiler.ts]: Try not to abbreviate variables too much, I don't think there is a lot of abbreviations on other variable names, we should probably try to be somewhat consistent.

File: src/TSHelper.ts (hunk 2 above)
Comment (tomblind) [reply src/TSHelper.ts]: lambda properties get a dummy context (versus stand-alone lambdas, which do not). This ensures they can be called the same as methods when obfuscated behind an interface.

**Comment by tomblind** [comment]
Variable names and the error message are updated. Sorry I was getting lazy there 😄

**Pull request merged**